django-stubs/db/models/fields/json.pyi

```diff
@@ -1,34 +1,100 @@
-        encoder: Optional[Any] = ...,
-        decoder: Optional[Any] = ...,
-        **kwargs: Any
-    ) -> None: ...
-    def check(self, **kwargs: Any) -> Any: ...
-    def deconstruct(self) -> Any: ...
+    encoder: json.JSONEncoder = ...
+    decoder: json.JSONEncoder = ...
     def from_db_value(self, value: Any, expression: Any, connection: Any) -> Any: ...
-    def get_internal_type(self) -> Any: ...
-    def get_prep_value(self, value: Any) -> Any: ...
     def get_transform(self, name: Any) -> Any: ...
-    def validate(self, value: Any, model_instance: Any) -> None: ...
     def value_to_string(self, obj: Any) -> Any: ...
-    def formfield(self, **kwargs: Any) -> Any: ...
+    @overload
+    def __init__(
+        self: JSONField[_A],
+        encoder: json.JSONEncoder = ...,
+        decoder: json.JSONDecoder = ...,
+        verbose_name: Optional[str] = ...,
+        name: Optional[str] = ...,
+        primary_key: bool = ...,
+        max_length: Optional[int] = ...,
+        unique: bool = ...,
+        blank: bool = ...,
+        null: Literal[False] = ...,
+        db_index: bool = ...,
+        default: Optional[Union[_A, Callable[[], _A]]] = ...,
+        editable: bool = ...,
+        auto_created: bool = ...,
+        serialize: bool = ...,
+        unique_for_date: Optional[str] = ...,
+        unique_for_month: Optional[str] = ...,
+        unique_for_year: Optional[str] = ...,
+        choices: Iterable[
+            Union[Tuple[_A, str], Tuple[str, Iterable[Tuple[_A, str]]]]
+        ] = ...,
+        help_text: str = ...,
+        db_column: Optional[str] = ...,
+        db_tablespace: Optional[str] = ...,
+        validators: Iterable[_ValidatorCallable] = ...,
+        error_messages: Optional[_ErrorMessagesToOverride] = ...,
+    ) -> None: ...
+    @overload
+    def __init__(
+        self: JSONField[Optional[_A]],
+        encoder: json.JSONEncoder = ...,
+        decoder: json.JSONDecoder = ...,
+        verbose_name: Optional[str] = ...,
+        name: Optional[str] = ...,
+        primary_key: bool = ...,
+        max_length: Optional[int] = ...,
+        unique: bool = ...,
+        blank: bool = ...,
+        null: Literal[True] = ...,
+        db_index: bool = ...,
+        default: Optional[Union[_A, Callable[[], _A]]] = ...,
+        editable: bool = ...,
+        auto_created: bool = ...,
+        serialize: bool = ...,
+        unique_for_date: Optional[str] = ...,
+        unique_for_month: Optional[str] = ...,
+        unique_for_year: Optional[str] = ...,
+        choices: Iterable[
+            Union[Tuple[_A, str], Tuple[str, Iterable[Tuple[_A, str]]]]
+        ] = ...,
+        help_text: str = ...,
+        db_column: Optional[str] = ...,
+        db_tablespace: Optional[str] = ...,
+        validators: Iterable[_ValidatorCallable] = ...,
+        error_messages: Optional[_ErrorMessagesToOverride] = ...,
+    ) -> None: ...
+    @overload
+    def __new__(  # type: ignore [misc]
```

Comment:
surprised we have both `__new__` and `__init__`, can we get away with just one of the two?

Reply:
I vaguely remember pyright preferring `__new__` overloads

Reply:
So, from what I understood from my debugging in there... Without `__new__` the `Literal[true/false]` for optional doesn't work (at least not for pyright).

The `__init__` is not mandatory, but it is used to validate the available options when creating the instance. E.g. without it, trying to do `foo = JSONField(encoder=<some_encoder>)` would give an error saying that there's no overload in `__init__` which supports encoder, because it would be using `Field.__init__`. We could set it to `__init__(*args, **kwargs)` but then we loose that validation, which also works for signature autocompletion on pyright.

And no, it doesn't use `__new__` for that, thus we have to have both, and more than one signature because of the `Literal[true/false]` thing. I would love to find a way to make this simpler though